### DISCUSSION TITLE
airsim msgpack rpc

### DISCUSSION THREAD

**Discussion by Flautdn** [microsoft/AirSim#4163] [Support Q&A]
Hi,
     In airsim client.py file the class VehicleClient has a constructor which defines the object client from msgpackrpc as

msgpackrpc.Client(msgpackrpc.Address(ip, port), timeout = timeout_value, pack_encoding = 'utf-8', unpack_encoding = 'utf-8') 

In it where is msgpackrpc server methods/classes defined in the airsim repository structure.

Explain the details on server files of Vehicle Client / Multirotor client server api.

Thanks.

**Comment by zimmy87** [comment]
Take a look at https://github.com/microsoft/AirSim/blob/master/AirLib/src/api/RpcLibServerBase.cpp and https://github.com/microsoft/AirSim/blob/master/AirLib/src/vehicles/multirotor/api/MultirotorRpcLibServer.cpp; these files contain the server bindings for the rpc server which determine which APIs can be called on the server. If you look in https://github.com/microsoft/AirSim/blob/master/PythonClient/airsim/client.py you'll see that most python client methods call `self.client.call()`. The parameter passed to that function corresponds to a binding in the server code.
